src/java_bytecode/java_static_initializers.cpp

```diff
@@ -135,82 +251,266 @@ static void create_clinit_wrapper_symbols(
+  static_assert((int) clinit_statest::NOT_INIT==0, "Check commment above");
+  static_assert((int) clinit_statest::IN_PROGRESS==1, "Check comment above");
+  static_assert((int) clinit_statest::INIT_COMPLETE==2, "Check comment above");
+#endif
+
+  // bool init_complete;
+  {
+    code_declt decl(init_complete.symbol_expr());
+    function_body.add(decl);
+  }
+
+  // if(C::__CPROVER_local_state == INIT_COMPLETE) return;
+  {
+    code_ifthenelset conditional;
+    conditional.cond() = gen_clinit_eqexpr(
+      clinit_thread_local_state_sym.symbol_expr(),
+      clinit_statest::INIT_COMPLETE);
+    conditional.then_case() = code_returnt();
+    function_body.add(conditional);
+  }
 
-  // the entire body of the function is an if-then-else
-  code_ifthenelset wrapper_body;
+  // C::__CPROVER_local_state = INIT_COMPLETE;
+  {
+    code_assignt assign = gen_clinit_assignexpr(
+      clinit_thread_local_state_sym.symbol_expr(),
+      clinit_statest::INIT_COMPLETE);
+    function_body.add(assign);
+  }
 
-  // add the condition to the if
-  wrapper_body.cond()=check_already_run;
+  // ATOMIC_BEGIN
+  {
+    function_body.add(atomic_begin);
+  }
 
-  // add the "already-run = false" statement
-  code_blockt init_body;
-  code_assignt set_already_run(already_run_symbol.symbol_expr(), true_exprt());
-  init_body.move_to_operands(set_already_run);
+  // Assume: clinit_state_sym != IN_PROGRESS
+  {
+    exprt assumption = gen_clinit_eqexpr(
+      clinit_state_sym.symbol_expr(), clinit_statest::IN_PROGRESS);
+    assumption.negate();
+    code_assumet assume(assumption);
+    function_body.add(assume);
+  }
 
-  // iterate through the base types and add recursive calls to the
-  // clinit_wrapper()
-  const symbolt &class_symbol = symbol_table.lookup_ref(class_name);
-  for(const auto &base : to_class_type(class_symbol.type).bases())
+  // If(C::__CPROVER_clinit_state == NOT_INIT)
+  // {
+  //   C::__CPROVER_clinit_state = IN_PROGRESS;
+  //   init_complete = false;
+  // }
+  // else If(C::__CPROVER_clinit_state == INIT_COMPLETE)
+  // {
+  //   init_complete=true;
+  // }
   {
-    const auto base_name = to_symbol_type(base.type()).get_identifier();
-    irep_idt base_init_routine = clinit_wrapper_name(base_name);
-    auto findit = symbol_table.symbols.find(base_init_routine);
-    if(findit == symbol_table.symbols.end())
-      continue;
-    code_function_callt call_base;
-    call_base.function() = findit->second.symbol_expr();
-    init_body.move_to_operands(call_base);
+    code_ifthenelset conditional;
+    code_blockt thenblk;
+    conditional.cond() = gen_clinit_eqexpr(
+      clinit_state_sym.symbol_expr(), clinit_statest::NOT_INIT);
+    thenblk.add(
+      gen_clinit_assignexpr(
+        clinit_state_sym.symbol_expr(), clinit_statest::IN_PROGRESS));
+    thenblk.add(code_assignt(init_complete.symbol_expr(), false_exprt()));
+    conditional.then_case() = thenblk;
+
+    code_ifthenelset condional_b;
```

Comment:
and this `init_conditional` to disambiguate these two blocks

Reply:
done